contracts/strategy/StrategyStrictBase.sol

```diff
@@ -0,0 +1,257 @@
+    }
+
+    uint amountAdjusted = Math.min(amount, balance);
+    if (amountAdjusted != 0) {
+      _decreaseBaseAmount(_asset, amountAdjusted);
+      IERC20(_asset).safeTransfer(_vault, amountAdjusted);
+    }
+    emit WithdrawToVault(amount, amountAdjusted, balance);
+  }
+
+
+  // *************************************************************
+  //                  baseAmounts modifications
+  // *************************************************************
+
+  /// @notice Decrease {baseAmounts} of the {asset} on {amount_}
+  ///         The {amount_} can be greater then total base amount value because it can includes rewards.
+  ///         We assume here, that base amounts are spent first, then rewards and any other profit-amounts
+  function _decreaseBaseAmount(address asset_, uint amount_) internal {
+    uint baseAmount = baseAmounts[asset_];
+    require(baseAmount >= amount_, WRONG_AMOUNT);
+    baseAmounts[asset_] = baseAmount - amount_;
+    emit UpdateBaseAmounts(asset_, -int(baseAmount));
+  }
+
+  /// @notice Increase {baseAmounts} of the {asset} on {amount_}, ensure that current {assetBalance_} >= {amount_}
+  /// @param assetBalance_ Current balance of the {asset} to check if {amount_} > the balance. Pass 0 to skip the check
+  function _increaseBaseAmount(address asset_, uint amount_, uint assetBalance_) internal {
+    baseAmounts[asset_] += amount_;
+    emit UpdateBaseAmounts(asset_, int(amount_));
+    require(assetBalance_ >= amount_, WRONG_AMOUNT);
+  }
+
+  // *************************************************************
+  //                       HELPERS
+  // *************************************************************
+
+  // todo: not sure if we need this function since priceChangeTolerance is always 0 for strict strategies.
+  /// @notice Calculate withdrawn amount in USD using the {assetPrice}.
+  ///         Revert if the amount is different from expected too much (high price impact)
+  /// @param balanceBefore Asset balance of the strategy before withdrawing
+  /// @param investedAssetsUSD Expected amount in USD, decimals are same to {_asset}
+  /// @param assetPrice Price of the asset, decimals 18
+  /// @return balance Current asset balance of the strategy
+  function _checkWithdrawImpact(
+    address _asset,
+    uint balanceBefore,
+    uint investedAssetsUSD,
+    uint assetPrice
+  ) internal view returns (uint balance) {
+    balance = IERC20(_asset).balanceOf(address(this));
+
+    if (assetPrice != 0 && investedAssetsUSD != 0) {
+      uint withdrew = balance > balanceBefore ? balance - balanceBefore : 0;
+      uint withdrewUSD = withdrew * assetPrice / 1e18;
+      uint priceChangeTolerance = 0;
+      uint difference = investedAssetsUSD > withdrewUSD ? investedAssetsUSD - withdrewUSD : 0;
+      require(difference * FEE_DENOMINATOR / investedAssetsUSD <= priceChangeTolerance, IMPACT_TOO_HIGH);
+    }
+  }
+
+  //todo: refactor think on permissions
+  function setVault(address _vault) external {
```

Comment:
We have a kind of circular dependency here. Vault needs a strategy and vice versa. I propose adding a such method to the vault to be able to deploy it w/o strategy. In addition, we need to add the StrictVault factory to configure all pieces properly. What do you think?

Reply:
just implement init pattern - anyone can setup the vault but only once

Reply:
Factory is good idea